### PR TITLE
feat: tree command utility

### LINKED ISSUE #1569: tree command to display all the supply chain security artifacts
**Description**

It'd be nice to have a tree command to display all the supply chain security artifacts such as SBOM, Attestation, Signature, etc. All of these are stored in an OCI registry along with the image itself with separated tags for SBOM .sbom, for Attestation .att, for Signature .sig, so, we'll be adding a tree command to view all of them in a nice formatted way like the following:

```shell
$ cosign tree <img> # if the user supplies a tag we'll be visualizing the data only for the tag, if not, we'll be summarizing all of the tags that the image has.

devopps/hello-world:latest
--> sha256:9fc31a8935dfdac754401cac99a6812a932fac70a49c003baecc2c2be0ba4454
    --> sha256-9fc31a8935dfdac754401cac99a6812a932fac70a49c003baecc2c2be0ba4454.sig
    --> sha256-9fc31a8935dfdac754401cac99a6812a932fac70a49c003baecc2c2be0ba4454.att
	--> sha256-9fc31a8935dfdac754401cac99a6812a932fac70a49c003baecc2c2be0ba4454.sbom

devopps/hello-world:v1
--> sha256:3267cad69a9cfc40c239f12c2589740642323772a445c837838a7f4d2507e605.
    --> sha256-3267cad69a9cfc40c239f12c2589740642323772a445c837838a7f4d2507e605.sig
    --> sha256-3267cad69a9cfc40c239f12c2589740642323772a445c837838a7f4d2507e605.att
	--> sha256-3267cad69a9cfc40c239f12c2589740642323772a445c837838a7f4d2507e605.sbom
```

### PR DESCRIPTION
**Pull request by developer-guy**
Fixes #1569

Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

Screen Shot 2022-03-14 at 12 14 11

#### Summary

This PR will display the Supply Chain Security related artifacts for an image including Signatures, Attestations and SBOMs

#### Ticket Link

Fixes #1569 

#### Release Note

```release-note
feat: tree command utility
```

### CODE REVIEW

File: cmd/cosign/cli/tree.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	if err != nil {
+		return err
+	}
+	_, err = remote.Get(sbomRef, remoteOpts...)
+	if err == nil {
+		scsaRefs = append(scsaRefs, sbomRef)
+	}
+
+	if len(scsaRefs) == 0 {
+		fmt.Fprintf(os.Stdout, "No SCSA artifacts found for image %s\n", ref.String())
```
Comment (hectorj2f) [inline cmd/cosign/cli/tree.go:94]: Perhaps it is better to avoid abbreviations here. Just prompt Supply Chain Security Related Artifacts as above.

Comment (developer-guy) [reply cmd/cosign/cli/tree.go]: Totally agree with you, it's an auto-completion issue. I've updated it according to your feedback, thanks.

File: cmd/cosign/cli/tree.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	sigRef, err := ociremote.SignatureTag(ref, ociremote.WithRemoteOptions(remoteOpts...))
+	if err != nil {
+		return err
+	}
+	_, err = remote.Get(sigRef, remoteOpts...)
+	if err == nil {
+		scsaRefs = append(scsaRefs, sigRef)
+	}
+
+	attRef, err := ociremote.AttestationTag(ref, ociremote.WithRemoteOptions(remoteOpts...))
```
Comment (hectorj2f) [inline cmd/cosign/cli/tree.go:75]: Would it work if you have multiple attestations?

**Review by hectorj2f — COMMENTED** [review]
Lgtm, i am not sure of the command's name. But i don’t have any alternative name.

File: cmd/cosign/cli/tree.go (hunk 2 above)
Comment (hectorj2f) [reply cmd/cosign/cli/tree.go:75]: Umm! When I run cosign download attestation I can see all my attestations. But the cosign tree command still show only one .sig and .att artifacts.

Comment (hectorj2f) [reply cmd/cosign/cli/tree.go:75]: I'd expect to see all my attestations and my signature.

Comment (developer-guy) [reply cmd/cosign/cli/tree.go:75]: What should the output of this command look like? What is your suggestion? Should we go one step further to list the layers of this tag?

Comment (hectorj2f) [reply cmd/cosign/cli/tree.go:75]: @developer-guy Generally, I'd say listing the layers would be idea. I was expecting to see something like:

```
cosign tree hectorj2f/golang:1.16
📦 Supply Chain Security Related artifacts for an image: hectorj2f/golang:1.16
│
├── 🔐 index.docker.io/hectorj2f/golang:sha256-953dfa4c0a09d03b197f8244a2c33e7fe0077e16add884dc91cf46f9bfe3e442.sig
├── 💾 index.docker.io/hectorj2f/golang:sha256-2e2af677fa07dd70d223fcfc376ccae2ec8e153429c04e60e236c68a1ac71a88.att
├── 💾 index.docker.io/hectorj2f/golang:sha256-bcd0cfee9996c5a8e3f175c8eeff5bd671b071684502f70493c3cdbbddc265b2.att
└── 📝 index.docker.io/hectorj2f/golang:sha256-953dfa4c0a09d03b197f8244a2c33e7fe0077e16add884dc91cf46f9bfe3e442.sbom
```

where I could see on my tree all my attestations. Let me know if you want me to add this code in a separate PR, or you prefer to take care of it.

Comment (developer-guy) [reply cmd/cosign/cli/tree.go:75]: Great, thank you, @hectorj2f. I'll take care of it ASAP.

**Review by hectorj2f — APPROVED** [review]

**Comment by hectorj2f** [comment]
@developer-guy It seems the codegen tests are failing https://github.com/sigstore/cosign/runs/5549954606?check_suite_focus=true.

**Comment by haydentherapper** [comment]
Is there an existing command this could go under, like `triangulate`? It seems unnecessary to me that this have a dedicated command.

**Comment by developer-guy** [comment]
It is more like a fun command to help people visualize every supply chain artifact related to an image in a unified way. Although, IMHO, the naming of that command is nice, triangulate only outputs the reference of an artifact, not visualize the whole tree, WDYT?

**Comment by haydentherapper** [comment]
They seem very related to me - Both commands are about providing more information about the OCI artifact.

**Pull request merged**